File: 1615.maximal-network-rank.go
/*
 * @lc app=leetcode id=1615 lang=golang
 *
 * [1615] Maximal Network Rank
 */

// @lc code=start
func maximalNetworkRank(n int, roads [][]int) int {
    roadIdsByCity := make([][]int, n);
	for id, road := range roads{
		roadIdsByCity[road[0]] = append(roadIdsByCity[road[0]], id)
		roadIdsByCity[road[1]] = append(roadIdsByCity[road[1]], id)
	}

	res := 0;
	for i := 0;i<n;i++{
		for j:=i+1;j<n;j++{
			set := make(map[string]bool)
			for _, roadId := range roadIdsByCity[i]{
				set[string(roadId)] = true;
			}
			for _, roadId := range roadIdsByCity[j]{
				set[string(roadId)] = true;
			}
			res = int(math.Max(float64(res), float64(len(set))))
		}
	}

	return res
}
// @lc code=end

File: 1615.maximal-network-rank.py
#
# @lc app=leetcode id=1615 lang=python3
#
# [1615] Maximal Network Rank
#

# @lc code=start
class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        roadIdsByCity = [[] for _ in range(n)]
        for city, road in enumerate(roads):
            roadIdsByCity[road[0]].append(city)
            roadIdsByCity[road[1]].append(city)
        
        res = 0
        for i in range(n):
            s = set()
            for roadId in roadIdsByCity[i]:
                s.add(roadId)
            for j in range(i+1, n):
                s_temp = s.copy()
                for roadId in roadIdsByCity[j]:
                    s_temp.add(roadId)
                res = max(res, len(s_temp))
        return res
# @lc code=end

